feat(EIP712): parsing of TypedData payload; encoding + hashing;

JeneaVranceanu opened the pull request:
## **Summary of Changes**

- Implementation of EIP712 TypedData payload parsing;
- Payload encoding & hashing.

Upd: this PR is focused on giving the ability to receive a JSON payload via `eth_signTypedData` RPC call, parse this payload, do some validations (though, validations can be improved for sure), encode message (a `message` is one of the fields of received JSON), hash message and according to given types (`types` also part of the received JSON and can be arbitrary), and prepare a special EIP712 hash that's used for signing of the received payload. 

In order to parse and sign this payload you as a user must do the following:
```swift
let eip712TypedData = try EIP712Parser.parse(receivedPayloadFromRPC)
let signature = try Web3Signer.signEIP712(
    eip712TypedData,
    keystore: keystore,
    account: account,
    password: password)
```

## **Test Data or Screenshots**

###### _By submitting this pull request, you are confirming the following:_

- I have reviewed the [Contribution Guidelines](https://github.com/web3swift-team/web3swift/blob/develop/CONTRIBUTION.md).
- I have performed a self-review of my own code.
- I have updated my repository to match the `develop` branch.
- I have included test data or screenshots that prove my fix is effective or that my feature works.
- I have checked that all tests work and swiftlint is not throwing any errors/warnings.

yaroslavyaroslav commented:
@JeneaVranceanu I wonder is it ready be reviewed now, or there would be some fixes added?

JeneaVranceanu commented:
> @JeneaVranceanu I wonder is it ready be reviewed now, or there would be some fixes added?

I think it makes sense to look into the old EIP712 implementation that we already had (and maybe even remove it 👀).

JeneaVranceanu commented:
I'll ping you by the end of this day with final thoughts/changes.

JeneaVranceanu commented:
@yaroslavyaroslav No more changes for this PR.
I've updated the documentation for EIP712Parser and the next steps will be:
- Check that `public class EIP712` covers all types possible;
- Updated `EIP712Domain` with all 5 properties according to the standard and add the ability to add more properties through a dictionary (maybe dictionary, not yet decided);
- Implement a parsing of custom statically typed objects into `EIP712TypedData`. As `EIP712TypedData` has all the encoding and hashing in place it will make sense to transform data from custom objects into `EIP712TypedData` and then use `encode`, `hash` or whatever is needed.

But all of that is for a different PR. This PR will have to wait for a while to get all these updates 🙂 
I suggest we review and merge this PR.

awaiskhan1304 commented:
Tested it with https://react-app.walletconnect.com/ and it is working just perfect.
Thanks @JeneaVranceanu.

yaroslavyaroslav commented:
> No more changes for this PR.

Started to reviewing it this Friday, considering to continue doing this today's evening. Quite tough workload I'm having now, sorry for a delay.

yaroslavyaroslav commented:
@JeneaVranceanu we have problems)

JeneaVranceanu commented:
> @JeneaVranceanu we have problems)

Yes, a failing test!
Added this morning. Have to implement support for custom type arrays.
And recursion support is missing.

JeneaVranceanu commented on Sources/Web3Core/EthereumABI/ABIParsing.swift:
```diff
@@ -7,16 +7,41 @@ import Foundation
-    public enum ParsingError: Swift.Error {
+    public enum ParsingError: LocalizedError {
         case invalidJsonFile
-        case elementTypeInvalid
+        case elementTypeInvalid(_ desc: String? = nil)
         case elementNameInvalid
         case functionInputInvalid
         case functionOutputInvalid
         case eventInputInvalid
         case parameterTypeInvalid
         case parameterTypeNotFound
         case abiInvalid
```
I've decided not to add `(_ desc: String? = nil)` too all error types as it would trigger more changes across the library.
It's better to do that updated in a separate PR.

JeneaVranceanu commented on Sources/web3swift/Utils/EIP/EIP712/EIP712.swift:
```diff
@@ -16,6 +16,12 @@ public class EIP712 {
     public typealias Bytes = Data
 }
 
+// FIXME: this type is wrong - The minimum number of optional fields is 5, and those are
+// string name the user readable name of signing domain, i.e. the name of the DApp or the protocol.
+// string version the current major version of the signing domain. Signatures from different versions are not compatible.
+// uint256 chainId the EIP-155 chain id. The user-agent should refuse signing if it does not match the currently active chain.
+// address verifyingContract the address of the contract that will verify the signature. The user-agent may do contract specific phishing prevention.
+// bytes32 salt an disambiguating salt for the protocol. This can be used as a domain separator of last resort.
 public struct EIP712Domain: EIP712Hashable {
```
This is something to deal with later.

JeneaVranceanu commented on Sources/Web3Core/KeystoreManager/BIP32Keystore.swift:
```diff
@@ -176,29 +176,29 @@ public class BIP32Keystore: AbstractKeystore {
         }
         guard let newNode = rootNode.derive(path: pathAppendix, derivePrivateKey: true) else {
-            throw AbstractKeystoreError.keyDerivationError
+            throw AbstractKeystoreError.keyDerivationError("BIP32Keystore. Failed to derive a new node. Check given parent node and path.")
```
We really need to revisit all our errors and make sure:
1) We have description;
2) The description makes sense to the developers who could get this error.

yaroslavyaroslav commented on Sources/Web3Core/KeystoreManager/BIP39.swift:
```diff
@@ -95,11 +95,13 @@ public class BIP39 {
         guard
-            size >= 128 && size <= 256 && size.isMultiple(of: 32),
-            let entropy = Data.randomBytes(length: size/8)
+            isCorrectSize,
```
Although it got better, it still doesn't fit with our convention of `guard else` formatting.

yaroslavyaroslav commented on Sources/Web3Core/KeystoreManager/EthereumKeystoreV3.swift:
```diff
@@ -23,13 +23,13 @@ public class EthereumKeystoreV3: AbstractKeystore {
-        if self.addresses?.count == 1 && account == self.addresses?.last {
-            guard let privateKey = try? self.getKeyData(password) else {
+        if account == addresses?.last {
+            guard let privateKey = try? getKeyData(password) else {
```
Maybe we can make this bit a bit more straightforward? Like changing the `if let` above to an another guard statement.

yaroslavyaroslav replied:
Upd: oh, I see that the execution flow continues after that line:32 throw statement. Maybe we should add at least `TODO: refactor me` at the very beginning of the whole file, as I believe this function should be split in to, but not this time?

yaroslavyaroslav commented on Sources/web3swift/Utils/EIP/EIP712/EIP712Parser.swift:
```diff
@@ -0,0 +1,297 @@
+}
+
+internal struct EIP712TypeArray: Codable {
+    public let types: [String : [EIP712TypeProperty]]
```
I wonder if this could be a concrete type rather the dictionary?

yaroslavyaroslav commented on Sources/web3swift/Utils/EIP/EIP712/EIP712Parser.swift:
```diff
@@ -0,0 +1,297 @@
+}
+
+public struct EIP712TypedData {
+    public let types: [String: [EIP712TypeProperty]]
```
Same question about the concrete type here.

yaroslavyaroslav commented on Sources/web3swift/Utils/EIP/EIP712/EIP712Parser.swift:
```diff
@@ -0,0 +1,297 @@
+        guard let primaryType = json["primaryType"] as? String else {
+            throw Web3Error.inputError(desc: "EIP712Parser. Top-level string field 'primaryType' missing.")
+        }
+        guard let domain = json["domain"] as? [String : AnyObject] else {
```
Maybe we should add here some at least short reasoning about why it's left here as a dictionary, rather being casted in a concrete type?

yaroslavyaroslav commented on Sources/web3swift/Utils/EIP/EIP712/EIP712Parser.swift:
```diff
@@ -0,0 +1,297 @@
+
+    /// Convenience function for ``encodeData(_:data:)`` that uses ``primaryType`` and ``message`` as values.
+    /// - Returns: encoded data based on ``primaryType`` and ``message``.
+    public func encodeData() throws -> Data {
```
I wonder why this is a one of a two public methods within the class that are having DocC comments? Shouldn't we add some least to other public one?

yaroslavyaroslav commented on Sources/web3swift/Utils/EIP/EIP712/EIP712Parser.swift:
```diff
@@ -0,0 +1,297 @@
+            throw Web3Error.processingError(desc: "EIP712Parser. Attempting to encode data for type that doesn't exist in this payload. Given type: \(type). Available types: \(types.keys).")
+        }
+
+        func encodeField(_ field: EIP712TypeProperty,
```
Why are we using approach of implementing one function within the other here?

yaroslavyaroslav commented on Sources/web3swift/Utils/EIP/EIP712/EIP712Parser.swift:
```diff
@@ -0,0 +1,297 @@
+                encValues.append(value as Any)
+            }
+            return (encTypes, encValues)
+        }
```
Would you please consider to split this one up to fourth separated yet serially called functioning sake of code clarity?

Those are: `213:220`, `222:226`, `239:256`, `256:263` and the last one the whole `if else` chaining.

yaroslavyaroslav commented on Sources/web3swift/Utils/Extensions/Data+Extension.swift:
```diff
@@ -0,0 +1,14 @@
+    func asJsonDictionary() throws -> [String: AnyObject]? {
+        try JSONSerialization.jsonObject(with: self, options: .mutableContainers) as? [String:AnyObject]
+    }
+}
```
We're still haven't added SwiftLint ruleset here, and I wonder would we ever do this, but I believe that such a set that we're having in a stash contains the rule about keeping one empty line at the very end of each file.

yaroslavyaroslav commented on Sources/web3swift/Utils/Extensions/String+Extension.swift:
```diff
@@ -0,0 +1,16 @@
+        guard let data = data(using: .utf8) else { return nil }
+        return try data.asJsonDictionary()
+    }
+}
```
We're still haven't added SwiftLint ruleset here, and I wonder would we ever do this, but I believe that such a set that we're having in a stash contains the rule about keeping one empty line at the very end of each file.

yaroslavyaroslav commented on Sources/web3swift/Web3/Web3+Signing.swift:
```diff
@@ -40,19 +40,36 @@ public struct Web3Signer {
+        guard let signature = try Web3Signer.signPersonalMessage(hash,
+                                                                 keystore: keystore,
+                                                                 account: account,
+                                                                 password: password ?? "",
```
Is this default password value necessary?

yaroslavyaroslav commented on Sources/web3swift/Web3/Web3+Signing.swift:
```diff
@@ -40,19 +40,36 @@ public struct Web3Signer {
                                                                  keystore: keystore,
                                                                  account: account,
-                                                                 password: password ?? "")
+                                                                 password: password ?? "",
```
Is this default password value necessary?

yaroslavyaroslav commented on Sources/Web3Core/EthereumABI/ABIParsing.swift:
```diff
@@ -7,16 +7,41 @@ import Foundation
+            case .invalidJsonFile:
+                errorMessage = ["invalidJsonFile"]
+            case .elementTypeInvalid(let desc):
+                errorMessage = ["elementTypeInvalid", desc]
```
I wonder if this line is a cause that we're messing here with an array of Strings rather a plain String?

yaroslavyaroslav commented on Sources/Web3Core/KeystoreManager/AbstractKeystore.swift:
```diff
@@ -11,11 +11,30 @@ public protocol AbstractKeystore {
+        var errorMessage: [String?]
+        switch self {
+        case .noEntropyError(let additionalDescription):
+            errorMessage = ["Entropy error (e.g. failed to generate a random array of bytes).", additionalDescription]
```
I wonder if this line is a cause that we're messing here with an array of Strings rather a plain String? Here it looks quite more necessary than in the previous alike code snippet, yet I wonder if it's a best way in terms of code intention clarity to implement such behaviour?
